### PR TITLE
Add `astro/types` for common prop patterns

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

- We have some common patterns that folks ask about, so this adds an `astro/types` entrypoint with some helpful utility types.
- This keeps folks from needing to access the `astroHTML.JSX.AnchorHTMLAttributes` stuff directly
- Based on docs section [TypeScript: Common Prop Type Patterns](https://docs.astro.build/en/guides/typescript/#common-prop-type-patterns)
## `HTMLAttributes`

If you would like to extend valid HTML attributes for a given HTML element, you may use the provided `HTMLAttributes` type—it accepts an element name and returns the valid HTML attributes for that element name.

```ts
import { HTMLAttributes } from 'astro/types';
interface Props extends HTMLAttributes<'a'> {
  myProp?: string;
};
```

## Testing

Tested manually

## Docs

https://github.com/withastro/docs/pull/1857

### CODE REVIEW

File: packages/astro/type-utils.d.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import './astro-jsx';
+import { AstroBuiltinAttributes } from './dist/@types/astro';
+
+/** Enforce that a component accepts no props or slots */
+export type NoProps = Record<string, never>;
+/** Enforce that a component requires a default slot */
+export type WithChildren = { children: any; };
```
Comment (Princesseuh) [inline packages/astro/type-utils.d.ts:7]: Having two different patterns here make this a bit weird for me, I also don't really like that `NoProps` implicitly means `NoChildren` (even though it does make total sense internally)

I'm struggling to think of a way to make this better though, hmm

File: .changeset/dirty-cycles-lick.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
+---
+'astro': minor
+---
+
+Add `astro/type-utils` entrypoint. These utilities can be used for common prop type patterns.
```
Comment (bholmesdev) [inline .changeset/dirty-cycles-lick.md:5]: Wondering if `type-utils` is the best name for this? I understand why these can't come from the `astro` global, but a simpler `astro/types` may be more flexible.

**Review by bholmesdev — COMMENTED** [review]
Have a few thoughts on naming and documentation. But `HTMLAttributes` feels fund`element`al to good Astro components 👏

File: .changeset/dirty-cycles-lick.md (hunk 2 above)
Comment (natemoo-re) [reply .changeset/dirty-cycles-lick.md:5]: Yeah I agree with that! `astro/types` seems better.

File: .changeset/dirty-cycles-lick.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+};
+```
+
+## `NoProps`
+
+If your component takes no props or slotted content, you can use the provided `NoProps` type.
+
+```ts
+import { NoProps } from 'astro/type-utils';
+type Props = NoProps;
```
[the remark this answers is not in the document]
Comment (natemoo-re) [reply .changeset/dirty-cycles-lick.md:24]: I think this is changing with the move to the compiler's TSX output, so maybe this type isn't needed. I'll remove it.

File: packages/astro/type-utils.d.ts (hunk 1 above)
Comment (natemoo-re) [reply packages/astro/type-utils.d.ts]: I don't know that these two will be useful in the long-term, so I'm going to reduce our API surface and just remove these until we have a stable plan for slots.

**Review by tony-sull — APPROVED** [review]
This is excellent!  Will be able to clean up the `@astrojs/image` components too, they're doing a bit of a dance to add typings when extending `` and `<picture>`

**Review by bholmesdev — APPROVED** [review]
That's my `type` of change!

**Pull request merged**